Sources/SwiftParser/StringLiterals.swift

```diff
@@ -44,6 +148,243 @@ extension Parser {
     }
   }
 
+  /// If `text` only consists of indentation whitespace (space and tab), return
+  /// the trivia pieces that represent `text`, otherwise return `nil`.
+  private func parseIndentationTrivia(text: SyntaxText) -> [RawTriviaPiece]? {
+    let trivia = TriviaParser.parseTrivia(text, position: .leading)
+    if trivia.allSatisfy({ $0.isIndentationWhitespace }) {
+      return trivia
+    } else {
+      return nil
+    }
+  }
+
+  private func postProcessMultilineStringLiteral(
+    openQuote: RawTokenSyntax,
+    segments allSegments: [RawStringLiteralSegmentsSyntax.Element],
+    closeQuote: RawTokenSyntax
+  ) -> (
+    unexpectedBeforeOpenQuote: [RawTokenSyntax],
+    openQuote: RawTokenSyntax,
+    segments: [RawStringLiteralSegmentsSyntax.Element],
+    unexpectedBeforeCloseQuote: [RawTokenSyntax],
+    closeQuote: RawTokenSyntax
+  ) {
+    // -------------------------------------------------------------------------
+    // Precondition
+
+    assert(openQuote.trailingTriviaByteLength == 0, "Open quote produced by the lexer should not have trailing trivia because we would drop it during post-processing")
+    assert(closeQuote.leadingTriviaByteLength == 0, "Closing quote produced by the lexer should not have leading trivia because we would drop it during post-processing")
+    assert(
+      allSegments.allSatisfy {
+        if case .stringSegment(let segment) = $0 {
+          return segment.unexpectedBeforeContent == nil
+            && segment.unexpectedAfterContent == nil
+            && segment.content.leadingTriviaByteLength == 0
+            && segment.content.trailingTriviaByteLength == 0
+        } else {
+          return true
+        }
+      },
+      "String segement produced by the lexer should not have unexpected text or trivia because we would drop it during post-processing"
+    )
+
+    // -------------------------------------------------------------------------
+    // Variables
+
+    var middleSegments = allSegments
```

Comment:
Seems like we could avoid removing from the start and then inserting later if we instead just kept separate arrays here, ie.
```
let multilineSegments = allSegments.dropFirst().dropLast() // this is a slice rather than the array
let expectedStart = allSegments.count > 1 ? allSegments.first : nil
let expectedIndentation = allSegments.last()

var processed = [RawStringLiteralSegmentsSyntax.Element]()
```

I can understand not wanting to do that though as it makes it a little easier to accidentally drop a segment, I still think it's cleaner but... up to you. I'd change the `removeLast` to just a `popLast` in either case (no `popFirst`, I assume because they didn't want to make that easy to do 🤷).

IMO it also wouldn't hurt to have a comment at the start of this function explaining the expected segments, ie. newline, the actual string segments, ending indentation + that no trivia is expected (the assertion does cover that though).

Reply:
Yeah, that one would have to be appended at the end :P. I was expecting this answer, so it's all good 👍